build/xtask/src/lsp.rs

```diff
@@ -0,0 +1,390 @@
+        //
+        // For example, if we are enabling features "foo" and "baz"
+        // ```toml
+        // foo = ["bar?/lol"]
+        // baz = ["bar"]
+        // ```
+        // when "foo" is checked, it does not know that "bar" is enabled.
+        //
+        // To work around this, we accumulate optional features in a separate
+        // list (`optional`), then recheck them after all of the mandatory
+        // features and dependencies have been handled.
+        //
+        // Once everything stabilizes, we know that any remaining optional
+        // features did not get enabled, and we break out of the loop.
+        loop {
+            let mut changed = false;
+            let mut optional = vec![];
+            while let Some((pkg_name, feat)) = todo.pop() {
+                // Anything not in `packages` is something from outside the
+                // workspace, so we don't care about it.
+                let pkg = match self.0.get(&pkg_name) {
+                    Some(pkg) => pkg,
+                    None => continue,
+                };
+
+                // If we've never seen this package before, then insert all of
+                // its non-optional dependencies with their features.
+                if !package_features.contains_key(&pkg_name) {
+                    assert!(!package_dependencies.contains_key(&pkg_name));
+
+                    // Start with no features enabled
+                    package_features.entry(pkg_name.clone()).or_default();
+                    package_dependencies.entry(pkg_name.clone()).or_default();
+                    changed = true;
+
+                    // Insert all non-optional, non-build dependencies
+                    for d in pkg.dependencies.iter() {
+                        if !d.optional && d.kind != DependencyKind::Build {
+                            // Record this dependency
+                            changed |= package_dependencies
+                                .get_mut(&pkg_name)
+                                .unwrap()
+                                .insert(d.name.clone());
+
+                            // Queue up the dependent package for evaluation
+                            todo.push((d.name.clone(), None));
+                            if d.uses_default_features {
+                                todo.push((
+                                    d.name.clone(),
+                                    Some("default".to_owned()),
+                                ));
+                            }
+                            for f in &d.features {
+                                todo.push((d.name.clone(), Some(f.to_owned())))
+                            }
+                        }
+                    }
+                }
+
+                // Check to see if we're also enabling a feature here
+                if feat.is_none() {
+                    continue;
+                }
+                let feat = feat.unwrap();
```

Comment:
Maybe

```rust
let Some(feat) = feat else {
    continue;
};
```

Reply:
Good idea, done in b609f504bcb05a04b55235342a3c8be0d44004ad